viam-cartographer/src/carto_facade_go/carto_facade.go

```diff
@@ -0,0 +1,278 @@
+func (vcl *CViamCartoLib) TerminateViamCartoLib() error {
+	status := C.viam_carto_lib_terminate((**C.viam_carto_lib)(unsafe.Pointer(vcl)))
+	if !successfulStatus(status) {
+		// TODO: how to make this error useful?
+		return errors.New(string(status))
+	}
+	return nil
+}
+
+// CViamCarto is a struct to hold the c type viam_carto
+type CViamCarto struct {
+	viamCarto *C.viam_carto
+}
+
+// TODO: properly pass config
+func getConfig() C.viam_carto_config {
+	vcc := C.viam_carto_config{}
+
+	sz := 1
+	pSensor := C.alloc_bstring_array(C.size_t(sz))
+	sensorSlice := unsafe.Slice(pSensor, sz)
+	sensorSlice[0] = C.bfromcstr(C.CString("rplidar"))
+
+	vcc.sensors = pSensor
+	vcc.sensors_len = C.int(sz)
+
+	dataDir := C.bfromcstr(C.CString("data_dir"))
+	vcc.data_dir = dataDir
+
+	componentReference := C.bfromcstr(C.CString("comp_ref"))
+	vcc.component_reference = componentReference
+
+	return vcc
+}
+
+func cleanConfig(vcc C.viam_carto_config) {
+	C.free_bstring_array(vcc.sensors)
+}
+
+// NewViamCarto calls viam_carto_init and returns a pointer to a viam carto object.
+func NewViamCarto(vcl CViamCartoLib) (CViamCarto, error) {
+	// TODO: get vcl as a param & initialize c queue channel here
+	pVc := C.alloc_viam_carto()
+	defer C.free_alloc_viam_carto(pVc)
+
+	ppVc := (**C.viam_carto)(unsafe.Pointer(pVc))
+
+	// TODO: Init with data
+	vcc := getConfig()
+	defer cleanConfig(vcc)
+	vcac := C.viam_carto_algo_config{}
+
+	status := C.viam_carto_init(ppVc, vcl.viamCartolib, vcc, vcac)
+	if !successfulStatus(status) {
+		// TODO: how to make this error useful?
+		return CViamCarto{}, fmt.Errorf("error initializing viam_carto status = %d", status)
+	}
+
+	vc := CViamCarto{viamCarto: (*ppVc)}
+	return vc, nil
+}
+
+// Start is a wrapper for viam_carto_start
+func (vc *CViamCarto) Start(ctx context.Context) error {
+	status := C.viam_carto_start(vc.viamCarto)
+
+	if !successfulStatus(status) {
+		// TODO: how to make this error useful?
+		return errors.New(string(status))
+	}
+
+	return nil
+}
+
+// Stop is a wrapper for viam_carto_stop
+func (vc *CViamCarto) Stop(ctx context.Context) error {
+	status := C.viam_carto_stop(vc.viamCarto)
+
+	if !successfulStatus(status) {
+		// TODO: how to make this error useful?
+		return errors.New(string(status))
+	}
+
+	return nil
+}
+
+// TerminateViamCarto calls viam_carto_terminate to clean up memory for viam carto
+func (vc *CViamCarto) TerminateViamCarto(ctx context.Context) error {
+	status := C.viam_carto_terminate(&vc.viamCarto)
+
+	if !successfulStatus(status) {
+		// TODO: how to make this error useful?
+		return errors.New(string(status))
+	}
+
+	return nil
+}
+
+// CViamCartoAddSensorReading is a struct to hold the c type viam_carto_sensor_reading
+type CViamCartoAddSensorReading struct {
+	vcasr C.viam_carto_sensor_reading
+}
+
+// AddSensorReading is a wrapper for viam_carto_add_sensor_reading
+func (vc *CViamCarto) AddSensorReading(ctx context.Context) error {
```

Comment:
Lets have this function actually take a lidar reading (probably a byte slice) & a timestamp & convert it into the appropriate C type.